packages/beacon-node/src/chain/chain.ts

```
@@ -401,12 +406,24 @@ export class BeaconChain implements IBeaconChain {
         return state && {state, executionOptimistic: isOptimisticBlock(block)};
       }
     } else {
-      // request for finalized state
+      return null;
+    }
+  }
+
+  async getHistoricalStateBySlot(slot: number): Promise<{state: Uint8Array; executionOptimistic: boolean} | null> {
+    const finalizedBlock = this.forkChoice.getFinalizedBlock();
 
-      // do not attempt regen, just check if state is already in DB
-      const state = await this.db.stateArchive.get(slot);
-      return state && {state, executionOptimistic: false};
+    if (slot >= finalizedBlock.slot) {
+      return null;
```

Would this `null` ever actually get reached?  Perhaps is better to throw an error here or something to spark us to check for invalid logic if it ever comes up.

---

preference for keeping this as a null, simplifies consumption of the APIs.